R/gips_class.R

```diff
@@ -1691,6 +1700,196 @@ print.summary.gips <- function(x, ...) {
+  }
+  
+  c(n0, edited_number_of_observations)
+}
+
+#' Extract Log-Likelihood for gips class
+#' 
+#' Calculates Log-Likelihood of the sample based on the `gips` object.
+#' 
+#' This will always be the biggest for `perm = "()"` (provided that `p <= n`).
+#' 
+#' If the found permutation will still require more parameters than `n`,
+#'     the Likelihood does not exist, thus the function returns `NULL`.
+#' 
+#' If the `projected_cov` (output of [project_matrix()])
+#'     is close to singular, the `NA` is returned.
+#' 
+#' @param object An object of class "gips"; usually a result of a [find_MAP()].
+#' @param ... Further arguments passed to or from other methods.
+#' @param tol A tolerance for `det(projected_cov)`.
+#'     If the det is smaller than `tol`, the `NA` is returned.
+#' 
+#' @section Calculation details:
+#' \eqn{L(`projected\_cov`) = \Pi_{k\in\{1,...,n\}} (}PDF of multivariate normal distribution with mean 0 and cov matrix `projected_cov` at point \eqn{Z_k)}
+#' 
+#' Let \eqn{`projected\_cov` = \pi_{\Gamma}(S)}.
+#' 
+#' Assume the mean was not estimated:
+#' 
+#' \eqn{ln(L(`projected\_cov`)) = - 0.5 * n * p * ln(2\pi) - 0.5 * n * ln(det(`projected\_cov`)) - 0.5 * n * p}
+#' 
+#' Assume the mean was estimated:
+#' 
+#' \eqn{ln(L(`projected\_cov`)) = - 0.5 * (n-1) * p * ln(2\pi) - 0.5 * (n-1) * ln(det(`projected\_cov`)) - 0.5 * (n-1) * p}
+#' 
+#' @importFrom stats logLik
+#' 
+#' @export
+#' 
+#' @seealso
+#' * [logLik()] - Generic function this [logLik.gips()] extends.
+#' * [find_MAP()] - Usually, the `logLik.gips()`
+#'     is called on the output of `find_MAP()`.
+#' * [AIC.gips()], [BIC.gips()] - Often, one is more
+#'     interested in an Information Criterion AIC or BIC.
+#' * [project_matrix()] - The function that can project
+#'     the known matrix of the found permutations space.
+#'     Mentioned in **Calculation details** section above
+#' 
+#' @examples 
+#' S <- matrix(c(5.15,2.05,3.60,1.99,
+#'               2.05,5.09,2.03,3.57,
+#'               3.60,2.03,5.21,1.97,
+#'               1.99,3.57,1.97,5.13), nrow = 4)
+#' g <- gips(S, 5)
+#' logLik(g) # -32.67048
+#' # For perm = "", which is default, there is p + choose(p, 2) degrees of freedom
+#' 
+#' g_map <- find_MAP(g, optimizer = "brute_force")
+#' logLik(g_map) # -32.6722 # this will always be smaller than `logLik(gips(S, n, perm = ""))`
+#' 
+#' g_n_too_small <- gips(S, 4)
+#' logLik(g_n_too_small) # NULL # the likelihood does not exists
+logLik.gips <- function(object, ..., tol = 1e-07){
+  validate_gips(object)
+  
+  original_cov <- attributes(object)[["S"]]
+  projected_cov <- project_matrix(original_cov, object[[1]])
+  p <- ncol(original_cov)
+  n <- attr(object, "number_of_observations")
+  
+  tmp <- get_n0_and_edited_number_of_observations_from_gips(object)
+  n0 <- tmp[1]
+  edited_number_of_observations <- tmp[2]
+  
+  if(n < n0){ # Likelihood is not defined in that setting
+    return(NULL)
```

Comment:
In this case, the naive estimate of covariance matrix ($Z^TZ$) is singular. Projected naive estimate as well. So raising NA here would be coherent with NA returned later in almost singular covariance matrix case.
Is diffrentiating between NULL and NA necessary?

Reply:
Thank You for your comment.

`NULL` is returned when the likelihood does not exist.
`NA` is returned when the likelihood does exist, but we cannot provide a satisfying estimation for it because the matrix is almost singular.

In aa4f22cad3b49cf2208a604e60365798c6c97b32 I have made an additional, different warning in `logic.gips` and a new documentation section **Existence of likelihood**. Do You think now it is explained well enough?

Reply:
I was wondering, whether making such a difference is relevant from user's and ours point of view. It's additional complexity with imo small gains.
But new explanations are welcome.

Reply:
I understand your uncertainty. This may not be very clear to have two different non-numerical outputs.

However, this could be way more common for the user to be in the situation of $n < n0$ than a near singular projected matrix. One could easily understand the model does exist for $n\ge n0$, so the `NULL` in this situation would be very surprising. When the `NA` is returned, this will indicate there was a different kind of problem occurring in the computations.

Reply:
Sorry for the confusion, but I tested that the R's calculations of log det are highly accurate. I could not produce an example matrix where the log det would be poorly calculated. My biggest bet was:

```
p <- 2000
my_result <- determinant(diag(1e-310, p))$modulus[1]
real_det <- p * log(10) * (-310)
real_det - my_result # 2.4e-08, so quite good approximation
```

The eigenvalues positive and smaller than `1e-310` would almost underflow to `0` (`1e-324` does), and `p > 500` is not a good idea anyway to use in the `gips` package.

However, considering the following:
```
determinant(diag(0, 4))$modulus[1] # -Inf
```
so it can overflow to `-Inf`. It can also overflow to `+Inf`, but only when the elements are itself `+Inf` (in my tests).

I decided to leave the warning for both types of overflow, but I will leave the `-Inf` as a result (because I think this is way more probable to happen and is a helpful output), so no `NA` will be produced by `logLik.gips()`.